jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -215,5 +696,1697 @@ public void clear()
+        public int get(byte[] bytes, int offset, int length)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("get array {} {}", this, length);
+            int got = 0;
+            for (Iterator<RetainableByteBuffer> i = _buffers.listIterator(); length > 0 && i.hasNext();)
+            {
+                RetainableByteBuffer buffer = i.next();
+                int l = buffer.get(bytes, offset, length);
+                got += l;
+                offset += l;
+                length -= l;
+
+                if (buffer.isEmpty())
+                {
+                    buffer.release();
+                    i.remove();
+                }
+            }
+            return got;
+        }
+
+        @Override
+        public boolean isDirect()
+        {
+            return _direct;
+        }
+
+        @Override
+        public boolean hasRemaining()
+        {
+            for (RetainableByteBuffer rbb : _buffers)
+                if (!rbb.isEmpty())
+                    return true;
+            return false;
+        }
+
+        @Override
+        public long skip(long length)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("skip {} {}", this, length);
+            long skipped = 0;
+            for (Iterator<RetainableByteBuffer> i = _buffers.listIterator(); length > 0 && i.hasNext();)
+            {
+                RetainableByteBuffer buffer = i.next();
+                long skip = buffer.skip(length);
+                skipped += skip;
+                length -= skip;
+
+                if (buffer.isEmpty())
+                {
+                    buffer.release();
+                    i.remove();
+                }
+            }
+            return skipped;
+        }
+
+        @Override
+        public void limit(long limit)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("limit {} {}", this, limit);
+            for (Iterator<RetainableByteBuffer> i = _buffers.iterator(); i.hasNext();)
+            {
+                RetainableByteBuffer buffer = i.next();
+
+                long size = buffer.size();
+                if (limit == 0)
+                {
+                    buffer.release();
+                    i.remove();
+                }
+                else if (limit < size)
+                {
+                    buffer.limit(limit);
+                    limit = 0;
+                }
+                else
+                {
+                    limit -= size;
+                }
+            }
+        }
+
+        @Override
+        public Mutable slice()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("slice {}", this);
+            List<RetainableByteBuffer> buffers = new ArrayList<>(_buffers.size());
+            for (RetainableByteBuffer rbb : _buffers)
+                buffers.add(rbb.slice());
+            return newSlice(buffers);
+        }
+
+        @Override
+        public Mutable slice(long length)
```

Comment:
This implementation of `slice()` does not retain, unlike the version it overrides.

Reply:
It doesn't need to. It retains the contained buffers and then returns an entirely new DynamicCapacity instance over those buffers.  Thus it implements the contract.  i.e it returns an independent set of pointers over shared data.   When the nested buffers are consumed from either the original or the slice, they are released, so when both original and slice are consumed (or explicitly released), then so are the nested buffers